pkg/util/clientauthorizer/admin.go

```diff
@@ -5,25 +5,88 @@ package clientauthorizer
 
 import (
 	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
 
 	"github.com/sirupsen/logrus"
 )
 
+// NewAdmin creates a new instance of ClientAuthorizer to be used with Admin API.
+// This authorizer allows connections only if they
+// contain a valid client certificate signed by `caBundlePath` and
+// the client certificate's CommonName equals `clientCertCommonName`.
+func NewAdmin(log *logrus.Entry, caBundlePath, clientCertCommonName string) ClientAuthorizer {
+	authorizer := &admin{
+		clientCertCommonName: clientCertCommonName,
+
+		log:      log,
+		readFile: ioutil.ReadFile,
+	}
+
+	authorizer.readCABundle(caBundlePath)
+	return authorizer
+}
+
 type admin struct {
-	log *logrus.Entry
+	roots                *x509.CertPool
+	clientCertCommonName string
+
+	log      *logrus.Entry
+	readFile func(filename string) ([]byte, error)
 }
 
-func NewAdmin(log *logrus.Entry) ClientAuthorizer {
-	return &admin{
-		log: log,
+func (a *admin) readCABundle(caBundlePath string) {
+	caBundle, err := a.readFile(caBundlePath)
+	if err != nil {
+		a.log.Error(err)
```

Comment:
If we can't read or decode ca cert - we log an error and return. The authorizer will be not ready witohut ca cert and declining the requests.

I'm not super happy about the fact that we do not return an error while creating the authorizer now (so we can fail on the startup). I didn't manage to come up with a good way of failing on the startup with the new way of instantiating `env`. An alternative would be to exit on the first call to `env.AdminClientAuthorizer()`, but I think it's a bad idea too: will make failures less obvious.

Any thoughts?

Reply:
I think that my suggestion to use sync.Once was a bad one because it hides the error.

Let's add another function to env called `InitializeAuthorizers() error` and call that early in cmd/aro/rp.go .  It calls NewAdmin which now returns (ClientAuthorizer, error).  It also calls NewARM but it's OK that this doesn't return any error right now.  You get the validation error early.  Anyone who tries to use an authorizer without having called the initialize will dereference a null pointer and die.  IsReady() can go back to returning true and the tests simplify again.